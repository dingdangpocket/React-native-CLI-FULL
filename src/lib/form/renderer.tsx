import { createElement, FC, memo, ReactNode } from 'react';
import { Text, View } from 'react-native';
import { FormTextInput, FormTextInputProps } from './elements/text';
import { FormData, FormElement } from './types';

interface Props<T extends FormData> {
  element: FormElement<T>;
  value: any;
  onChange: (values: Partial<T>) => void;
}

const TextContent = memo(({ text }: { text?: ReactNode }) => {
  if (!text) return null;
  if (
    typeof text === 'string' ||
    typeof text === 'number' ||
    typeof text === 'boolean'
  ) {
    return <Text>{String(text)}</Text>;
  }
  return <View>{text}</View>;
});

const FormElementLabel = memo(({ label }: { label?: ReactNode }) => {
  return <TextContent text={label} />;
});

const FormHelpText = memo(({ text }: { text?: ReactNode }) => {
  return <TextContent text={text} />;
});

export const FormElementRenderer = memo(
  <T extends FormData>({ element, value, onChange }: Props<T>) => {
    const { label, labelPosition, helpText, ...props } = element;
    const orientation = label && labelPosition === 'left' ? 'h' : 'v';

    return (
      <View
        css={`
          flex-direction: ${orientation === 'h' ? 'row' : 'column'};
          flex-wrap: nowrap;
          justify-content: ${orientation === 'h'
            ? 'space-between'
            : 'flex-start'};
          align-items: ${orientation === 'h' ? 'center' : 'flex-start'};
        `}
      >
        <FormElementLabel label={label} />
        <View>
          {(() => {
            switch (element.type) {
              case 'text': {
                return createElement(
                  FormTextInput as unknown as FC<FormTextInputProps<T>>,
                  {
                    ...props,
                    value,
                    onChange,
                  },
                );
              }
              default:
                return null;
            }
          })()}
        </View>
        <FormHelpText text={helpText} />
      </View>
    );
  },
);
