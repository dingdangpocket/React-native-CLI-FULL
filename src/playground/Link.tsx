import { ReactNode, useCallback } from 'react';

export const Link = ({
  page,
  children,
}: {
  page: string;
  children: ReactNode;
}) => {
  const onClick = useCallback(() => {
    console.log(`click link ${page}`);
  }, [page]);

  return (
    <a href={page} className="normal" onClick={onClick}>
      {children}
    </a>
  );
};
