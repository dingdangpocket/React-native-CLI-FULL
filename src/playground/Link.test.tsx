/**
 * @file: Link.test.ts
 * @copyright: (c) 2019-2021 sichuan zhichetech co., ltd.
 */

import renderer from 'react-test-renderer';
import { Link } from './Link';

it('renders correctly', () => {
  const tree = renderer
    .create(<Link page="https://facebook.com">Facebook.com</Link>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
