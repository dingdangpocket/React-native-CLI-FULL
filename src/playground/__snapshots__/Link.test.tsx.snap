// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<a
  className="normal"
  href="https://facebook.com"
  onClick={[Function]}
>
  Facebook.com
</a>
`;
